### PR TITLE
SONARJAVA-4226: Use incremental analysis with caching for S4032

### PR DESCRIPTION
**Pull request by johann-beleites-sonarsource**

### CODE REVIEW

File: java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java [hunk 1]
```diff
@@ -19,32 +19,68 @@
...
       .onFile(testSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFiles/package-info.java"))
       .withCheck(new UselessPackageInfoCheck())
       .verifyIssueOnFile("Remove this package.");
   }
 
   @Test
   void withOtherFile() {
-    CheckVerifier.newVerifier()
+    verifier
       .onFile(testSourcesPath("checks/UselessPackageInfoCheck/package-info.java"))
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java:76]:
```suggestion
      .onFile(mainCodeSourcesPath("checks/UselessPackageInfoCheck/package-info.java"))
```

File: java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java [hunk 2]
```diff
@@ -19,32 +19,68 @@
...
+    this.writeCache = new InternalWriteCache().bind(readCache);
+    this.verifier = CheckVerifier.newVerifier()
+      .withCache(readCache, writeCache);
+  }
+
   @Test
   void withNoOtherFile() {
-    CheckVerifier.newVerifier()
+    verifier
       .onFile(testSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFiles/package-info.java"))
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java:68]:
```suggestion
      .onFile(mainCodeSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFiles/package-info.java"))
```

File: java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java [hunk 3]
```diff
@@ -19,32 +19,68 @@
...
       .verifyNoIssues();
   }
 
   @Test
   void notAPackageInfo() {
-    CheckVerifier.newVerifier()
+    verifier
       .onFiles(
         testSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFilesButNotPackageInfo/HelloWorld1.java"),
         testSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFilesButNotPackageInfo/HelloWorld2.java"))
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java:86]:
```suggestion
        mainCodeSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFilesButNotPackageInfo/HelloWorld1.java"),
        mainCodeSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFilesButNotPackageInfo/HelloWorld2.java"))
```

File: java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java [hunk 4]
```diff
@@ -54,18 +90,100 @@ void notAPackageInfo() {
 
   @Test
   void notAPackageInfoOnSingleFile() {
-    CheckVerifier.newVerifier()
+    verifier
       .onFile(testSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFilesButNotPackageInfo/HelloWorld1.java"))
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java:94]:
```suggestion
      .onFile(mainCodeSourcesPath("checks/UselessPackageInfoCheck/packageWithNoOtherFilesButNotPackageInfo/HelloWorld1.java"))
```

File: java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java [hunk 5]
```diff
@@ -54,18 +90,100 @@ void notAPackageInfo() {
...
       .withCheck(new UselessPackageInfoCheck())
       .verifyNoIssues();
   }
 
   @Test
   void defaultPackage() {
-    CheckVerifier.newVerifier()
+    verifier
       .onFile(testSourcesPath("DefaultPackage.java"))
```
Comment (dorian-burihabwa-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/UselessPackageInfoCheckTest.java:102]:
```suggestion
      .onFile(mainCodeSourcesPath("DefaultPackage.java"))
```

**Review by dorian-burihabwa-sonarsource — CHANGES_REQUESTED** [review]
Looks good to me 👍🏿  a few changes would help clarify the expectations.

File: java-checks-testkit/src/test/java/org/sonar/java/checks/verifier/internal/InternalReadCacheTest.java [hunk 6]
```diff
@@ -24,7 +24,9 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
+import org.sonar.api.batch.sensor.cache.ReadCache;
 
+import static org.assertj.core.api.Assertions.as;
```
[the remark this answers is not in the document]
Comment (johann-beleites-sonarsource) [reply java-checks-testkit/src/test/java/org/sonar/java/checks/verifier/internal/InternalReadCacheTest.java:29]: Not sure how the `as` import got in there - we do need the `ReadCache` one, as we implement it.

**Pull request merged**